### ISSUE TITLE
Publish from MessageReceivedHandler doesn't work

### ISSUE THREAD

**Issue by AlexShovkoplyas** [chkr1011/MQTTnet#648]
When I Publish message from handler

`client.UseApplicationMessageReceivedHandler(async e =>
 {
      await client.PublishAsync(message);
 });`

 I am getting MqttCommunicationTimedOutException from MqttPacketAwaiter before receiving acknowledgement that message was Published.

Despite I have set very long communication timeout - it is throwed much earlier.
I am using QOS 1.

Workaround that we've currently decided to use - to publish asynchronously:

`client.UseApplicationMessageReceivedHandler(e =>
 {
      Task.Run(async () => await client.PublishAsync(message));
 });`

**Comment by ramonsmits** [comment]
Seems similar to https://github.com/chkr1011/MQTTnet/issues/587

**Comment by ramonsmits** [comment]
@AlexShovkoplyas In my issue I couldn't create a very small reproduction. Can you?

**Comment by AlexShovkoplyas** [comment]
Sure, I've made very simple solution, please look: https://github.com/AlexShovkoplyas/MqttNetBug

I have added logging for clear debugging.

I've spent 2 days on debugging, I have a feeling that I understand solution and flow, see where exception is thrown, but I can't find source of this error and how to solve it.

**Comment by chkr1011** [comment]
Thank you for building the example. There is no way to fix this. You have to start a new task if you want to send messages inside the handler. This was introduced in 3.0.0. The goal is to preserve the order of the messages. So the client will wait for message processing before processing the next. The problem is that sending a message while waiting at the same time leads to a dead lock. 

You can now:
1. Use Task.Run to process handling.
2. Create a queue and use a worker tasks to process messages (but this will accept the message even before the real processing happened).

I will have a look if there is a way to change this behavior but starting new tasks internally is a overhead which is also not wanted on the other hand.

**Comment by AlexShovkoplyas** [comment]
Many thanks for a rich response!
I am wondering is it a limitation of protocol or client implementation.

```
mqttClient.UseApplicationMessageReceivedHandler(async e =>
   {
        Task.Run(async () => await mqttClient.PublishAsync(messageMqtt2));
        await Task.Delay(2000);
   });
```
This code runs and brings reshufled ordering of actions :
1. Receiving message (1)
2. Sending message (2)
3. Receiving acknowledgement (2)
4. Sending acknowledgement (1)

**Comment by AlexShovkoplyas** [comment]
As I guess action 1. blocks thread and It can't proceed with action 3. before finishing action 4.
BTW approach with Task.Run() can violate ordering of new published messages (brings a limitation)

**Comment by ramonsmits** [comment]
@AlexShovkoplyas you still much await the `Task.Run` other wise the handler could complete without having certainty that the Publish was successful which usually is of importance when processing at-least-once.

@chkr1011 Do I understand correctly that this means that using a second client instance would resolve this issue? Meaning, a client instance specifically for publishing?

**Comment by AlexShovkoplyas** [comment]
@ramonsmits I'am not sure if I understood you correctly but 

```
mqttClient.UseApplicationMessageReceivedHandler(async e =>
   {
        await Task.Run(async () => await mqttClient.PublishAsync(messageMqtt2));
   });
```
doesn't work..

**Comment by ramonsmits** [comment]
@AlexShovkoplyas Yes, it will result in the same issue. IMHO when publishing within an handler on the same client this should result in an exception as this is unsupported as @chkr1011 states.

Removing `await` will result in an at-most-once processing model as it is not guaranteed the handler will process successfully.

**Comment by AlexShovkoplyas** [comment]
Nice idea with a second client.
```
mqttClient.UseApplicationMessageReceivedHandler(async e =>
   {
        await mqttClient2.PublishAsync(message);
   });
```
works.. 
I will investigate if there are any limitations.

**Comment by chkr1011** [comment]
You can use a second client if that is OK in your UseCase (new connection, new client ID etc.). The problem is that there are different use cases out there. If you specify that a message is only processed successfully if a another message was delivered (including ACK) then there is a dead lock. 

I personally have a separate queue in my apps (BlockingCollection) and a different worker thread. So I am never blocked but I always send ACK even if the real operation was not performed yet. To detect that I have some more topics or payloads which contain some custom data which indicates a success.

I also thought about adding an internal queue for PUBLISH packets so that other packages like some ACK will be still processed but the problem is that this queue will grow and grow and maybe the receive speed is higher than processing speed so that the queue will grow forever. Adding a limit for that queue may lead to the same dead lock at some point.

**Comment by AlexShovkoplyas** [comment]
Thank you for support. Just would like to share feedback. 
We have stand on solution where we have client's wrapper which encapsulates 2 mqtt clients : 1 publisher and 1 subscriber.
Solution with queue also seems attractive, if message processing takes more time it would be easy for us to refactor and add queue layer later but for now we were concerned about durability of messages.

**Comment by ramonsmits** [comment]
> I personally have a separate queue in my apps (BlockingCollection) and a different worker thread. So I am never blocked but I always send ACK even if the real operation was not performed yet. 

That is fine for unreliable scenarios where at-most-once is allowed and not being able to recovery from errors is fine too but that does not seem applicable for most environments, only for simple things like notifications that will not trigger any business logic.

**Comment by chkr1011** [comment]
From MQTT spec:

> 1 The receiver is not required to complete delivery of the Application Message before sending the PUBACK. When its original sender receives the PUBACK packet, ownership of the Application Message is transferred to the receiver.

It just mentioned "ownership". So it does not include a full blown business logic. This makes sense to me because we are talking about a telemetry transport protocol only. Not a full blown RPC protocol like SOAP etc.
